courses/utils.py

```diff
@@ -0,0 +1,220 @@
+"""
+Utilities for courses/certificates
+"""
+import logging
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured, PermissionDenied
+from edx_api.client import EdxApi
+
+
+from courses.models import CourseRun, CourseRunGrade, CourseRunCertificate
+from users.api import User, fetch_user
+
+logger = logging.getLogger(__name__)
+
+
+def _verify_configuration():
+    """
+    Verify if required variables are setup properly
+    """
+    if settings.EDX_ACCESS_TOKEN is None:
+        raise ImproperlyConfigured("EDX_ACCESS_TOKEN is not set")
+
+    if settings.OPENEDX_API_KEY is None:
+        raise ImproperlyConfigured("OPENEDX_API_KEY is not set")
+
+
+def _get_course_run(readable_id):
+    """
+    Get the course run needed
+    """
+    course_run = CourseRun.objects.filter(courseware_id=readable_id).first()
+
+    if not course_run:
+        exception = CourseRun.DoesNotExist(
+            "Could not find course run with courseware_id={}".format(readable_id)
+        )
+        logger.error(str(exception))
+        raise exception
+    return course_run
+
+
+def _generate_grade(user, course_run, edx_grade, force=False):
+    """
+    Generate course run grade for a user
+    """
+    if (
+        CourseRunGrade.objects.filter(user=user, course_run=course_run).exists()
+        and not force
+    ):
+        logger.warning(
+            "Grade exists for user: %s and courseware_id: %s. Not overwriting",
+            user.email,
+            course_run.courseware_id,
+        )
+    else:
+        CourseRunGrade.objects.update_or_create(
+            user=user,
+            course_run=course_run,
+            defaults={
+                "grade": edx_grade.percent,
+                "passed": edx_grade.passed,
+                "letter_grade": edx_grade.letter_grade,
+            },
+        )
+
+
+def sync_courserun_grades(readable_id, email=None, override_grade=None, force=False):
+    """
+    Synchronize grades from OpenEdX to xPRO database
+    """
+    _verify_configuration()
+
+    edx_client = EdxApi(
+        {
+            "access_token": settings.EDX_ACCESS_TOKEN,
+            "api_key": settings.OPENEDX_API_KEY,
+        },
+        settings.OPENEDX_API_BASE_URL,
+    )
+    user = None
+    if email:
+        user = fetch_user(email)
+
+    course_run = _get_course_run(readable_id)
+
+    # Grade can also be 0.0 so `if not grade` does not work here
+    # We only sync grades if the grade is not provided
+    if override_grade is None:
+        # Grade synced from edX for a specific user for the course run
+        if user:
+            edx_grade = edx_client.current_grades.get_student_current_grade(
```

Comment:
any particular check for `edx_grade` before calling `_generate_grade` ?

Reply:
If the user is not enrolled the API already raises a 404, which in turn ends the command process. Otherwise it will be a valid `edx_api.grades.CurrentGrade` object.

Reply:
I've added a `try:except:else` anyway